### PR TITLE
task/RFR-13_transform-measurement-handler-to-kotlin

### PR DESCRIPTION
**Pull request by muthenberg**
This also updates the Gradle build files to allow custom Kotlin Code

### CODE REVIEW

File: executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.util.*
+import java.util.stream.Collectors
+
+/**
+ * A handler for receiving HTTP PUT requests on the "measurements" end point.
+ * This end point is the core of this application and responsible for receiving
+ * new measurements from any measurement device and forwarding those
+ * measurements for persistent storage.
```
Comment (hb0) [inline executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt:52]: Does not seem like our default line break length but shorter.

File: executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
+    init {
+        Validate.notNull(mongoClient)
+        Validate.isTrue(payloadLimit > 0)
+        this.mongoClient = mongoClient
+        this.payloadLimit = payloadLimit
+    }
+
+    override fun handleAuthorizedRequest(
+        ctx: RoutingContext, users: Set<User>,
+        header: MultiMap
```
Comment (hb0) [inline executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt:95]: weird format: ether each parameter in a new line or all 3 in one line, as they are short enough

File: executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+            val fs = ctx.vertx().fileSystem()
+            val path = session.get<Any>(UPLOAD_PATH_FIELD) as String
+            request.pause()
+            fs.exists(path).onSuccess { fileExists: Boolean? ->
+                try {
+                    request.resume()
+                    if (!fileExists!!) {
+                        session.remove<Any>(UPLOAD_PATH_FIELD) // was linked to non-existing file
+                        handleFirstChunkUpload(ctx, request, session, loggedInUser, contentRange, metaData, path)
+                        return@onSuccess
```
Comment (hb0) [inline executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt:133]: Just out of interest, was means this `@onSuccess` ? Just documentation?

File: executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt [hunk 4]
```diff
@@ -0,0 +1,573 @@
...
+                // Can't return 308. Google API client lib throws `Preconditions` on subsequent chunk upload.
+                // This makes sense as the server reported before that bytes (>0) were received.
+                LOGGER.warn(String.format("Response: 404, Unexpected content range: %s", contentRange))
+            }
+            ctx.response().setStatusCode(NOT_FOUND).end() // client sends a new pre-request for this upload
+            return
+        }
+        acceptUpload(ctx, request, session, user, contentRange, metaData)
+    }
+    // Cache found, expecting upload to continue the cached file
```
Comment (hb0) [inline executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt:199]: This comment got lost. Please find its previous home.

Comment (hb0) [reply executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt:199]: Ah, probably at line 201?

File: executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+                    bodySize, contentRange
+                )
+            )
+        }
+        return contentRange
+    }
+
+    /**
+     * Creates a new upload temp file and calls
+     * [.acceptUpload].
```
Comment (hb0) [inline executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt:345]: strange line break

**Review by hb0 — CHANGES_REQUESTED** [review]
See change requests

File: executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt (hunk 1 above)
Comment (muthenberg) [reply executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt:52]: Yes. This is the default Kotlin Style.

File: executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt (hunk 3 above)
Comment (muthenberg) [reply executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt:133]: I have no idea. This code was just converted from Java automatically by IntelliJ. I only checked that it compiles and builds. I'll have to rip out almost everything from here anyways for the new Implementation with Picture support, so I did not bother to fiddle around with this stuff too much.

File: executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt (hunk 4 above)
Comment (muthenberg) [reply executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt:199]: This is in exactly the same position as in the old Java file, namely above the documentation for the method `handleSubsequentChunkUpload`

File: executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt (hunk 5 above)
Comment (muthenberg) [reply executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt:345]: This is how a link is formatted in Kotlin. In java this includes all the parameters and thus requires a new line. The converter just took this over.

File: executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+import io.vertx.ext.mongo.MongoClient
+import io.vertx.ext.mongo.MongoGridFsClient
+import io.vertx.ext.web.RoutingContext
+import io.vertx.ext.web.Session
+import org.apache.commons.lang3.Validate
+import org.slf4j.LoggerFactory
+import java.io.File
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.util.*
```
[the remark this answers is not in the document]
Comment (hb0) [reply executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt]: I added this to our dev intro:
https://cyface.atlassian.net/wiki/spaces/IM/pages/26602177/Entwickler-Einstellungen#Whildcast-Imports-deaktivieren

File: executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt (hunk 5 above)
Comment (hb0) [reply executables/collector/src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt]: I see. Thanks for clarification, I have to get used to the Kotlin world I guess.

**Review by hb0 — DISMISSED** [review]

**Review by hb0 — APPROVED** [review]

**Pull request merged**